lib/coap/transport.ex

```diff
@@ -0,0 +1,432 @@
+    {:keep_state, s, {:state_timeout, s.retransmit_timeout, m}}
+  end
+
+  #
+  # STATE: :ack_pending
+  #
+
+  # M_CMD[ACK]
+  def handle_event(:info, %Message{type: :ack, message_id: mid} = m, {:ack_pending, mid}, s) do
+    tx(s, m)
+    {:next_state, :closed, s}
+  end
+
+  # M_CMD[*]
+  def handle_event(:info, %Message{}, {:ack_pending, _}, _s),
+    do: {:keep_state_and_data, :postpone}
+
+  @impl GenStateMachine
+  def terminate(_reason, _state, %{socket: nil}),
+    do: :ok
+
+  def terminate(_reason, _state, s) do
+    send(s.socket, :close)
+  end
+
+  ###
+  ### Priv
+  ###
+  defp when_valid?(%__MODULE__{error: nil} = state, fun),
+    do: fun.(state)
+
+  defp when_valid?(s, _), do: s
+
+  defp init_state(s, args) do
+    options =
+      args
+      |> Map.take([
+        :ack_timeout,
+        :ack_random_factor,
+        :max_retransmit,
+        :transport_opts,
+        :socket_init
+      ])
+
+    s
+    |> Map.merge(options, fn
+      _key, v1, nil -> v1
+      _key, _v1, v2 -> v2
+    end)
+    |> fill_retransmit_timeout()
+    |> validate_ack_random_factor()
+  end
+
+  defp cast_peer(s, %{peer: {{a, b, c, d}, port}}) do
+    uri = %URI{scheme: "coap", host: "#{a}.#{b}.#{c}.#{d}", port: port}
+    %{s | peer: uri}
+  end
+
+  defp cast_peer(s, %{peer: {host, port}}) do
+    uri = %URI{scheme: "coap", host: "#{host}", port: port}
+    %{s | peer: uri}
+  end
+
+  defp cast_peer(s, _args),
+    do: s
+
+  defp cast_socket_init(%{peer: %URI{scheme: "coap"}} = s),
+    do: %{s | socket_init: &CoAP.Transport.UDP.start/1}
+
+  defp cast_socket_init(%{socket_init: init} = s) when is_function(init, 1),
+    do: s
+
+  defp cast_socket_init(s) do
+    %{s | error: {:badarg, "Missing :peer or :socket_init"}}
+  end
+
+  defp open_socket(%{socket_init: init, peer: uri, transport_opts: opts} = s) do
+    case init.({uri, self(), opts}) do
+      {:ok, pid} ->
+        ref = Process.monitor(pid)
+        %{s | socket: pid, socket_ref: ref}
+
+      {:error, reason} ->
+        %{s | error: reason}
+    end
+  end
+
+  defp fill_retransmit_timeout(s) do
+    %{s | retransmit_timeout: __retransmit_timeout__(s.ack_timeout, s.ack_random_factor)}
+  end
+
+  defp validate_ack_random_factor(%{ack_random_factor: factor} = s)
+       when is_float(factor) and factor >= 1.0,
+       do: s
+
+  defp validate_ack_random_factor(s),
+    do: %{s | error: {:badarg, :ack_random_factor, s.ack_random_factor}}
+
+  def __max_transmit_wait__(ack_timeout, max_retransmit, ack_random_factor \\ @ack_random_factor) do
+    round(ack_timeout * (:math.pow(2, max_retransmit + 1) - 1) * ack_random_factor)
+  end
+
+  def __retransmit_timeout__(ack_timeout, ack_random_factor \\ @ack_random_factor) do
+    random = :rand.uniform() * ack_timeout * (ack_random_factor - 1)
+    round(ack_timeout + random)
+  end
+
+  defp tx(%__MODULE__{socket: socket}, m) do
+    send(socket, {:send, m})
+  end
+
+  defp rr_evt(%__MODULE__{client: client}, evt) do
```

Comment:
Do you mind if we rename function this to the full name? I see a lot of shortening that makes this feel like it's crossing a threshold into making it a little difficult to follow. `s` as `state` makes sense, but it feels like we maybe cross a line in a few spots.

Reply:
Changes names:
* tx -> tx_socket
* rr_evt -> request_response_event